packages/sdk/src/oracle/store/write.ts

```diff
@@ -0,0 +1,107 @@
+import type * as ethers from "../types/ethers";
+import type * as state from "../types/state";
+
+// This file contains composable and type safe state writers which mirror the state in types/state.
+// Each component takes in 1 parameters, state and you can include any number of functions to operate on the state.
+// Some things to consider:
+// 1. State can be nested, but it should strictly one direction, each class should ideally only operate on its direct state object.
+// 2. The parent can return new components, but its responsible for initializing the state for the child.
+// 3. You can modify this.state in the component thanks to immer, but you cannot set this.state to a new object, only its properties.
+
+export class User {
+  constructor(private state: Partial<state.User>) {}
+  set(data: Partial<state.User>): void {
+    // note that this is done because we cannot replace this.state = data or immer loses visibility to the change.
+    if (data.chainId) this.chainId(data.chainId);
+    if (data.address) this.address(data.address);
+    if (data.signer) this.signer(data.signer);
+  }
+  chainId(chainId: number): void {
+    this.state.chainId = chainId;
+  }
+  address(address: string): void {
+    this.state.address = address;
+  }
+  signer(signer: ethers.Signer): void {
+    this.state.signer = signer;
+  }
+}
+export class Balances {
+  constructor(private state: Partial<state.Balances>) {}
+  set(address: string, amount: ethers.BigNumber): void {
+    this.state[address] = amount;
+  }
+}
+export class Erc20 {
+  constructor(private state: Partial<state.Erc20>) {}
+  props(data: state.Erc20["props"]): void {
+    this.state.props = data;
+  }
+  balance(account: string, amount: ethers.BigNumber): void {
+    if (!this.state.balances) this.state.balances = {};
+    new Balances(this.state.balances).set(account, amount);
```

Comment:
nit: what's the point of this ephemeral balances class that's only used to set a value in a mapping? Why not store or return the balances class?

Reply:
The balance class is used for setting both balances and allowances. its also reused for allowances.

this balance class may be extended eventually to not just set an amount, but add or remove. In that case we would probably change this to return the balances class, so you could have like `balance.add(address,amount)`, `balance.sub(address,amount)`. Right now theres just a single call to `set` so i figured it could save lines by just doing the set in the function without having to return the class